File: csharp/DesiArchitect.DesignPatterns/04_Decorator/Before/BillCalculator.cs
namespace DesiArchitect.DesignPatterns.Decorator.Before;

// Boolean flags for each charge; every new fee adds another if-block
public class BillCalculator
{
    public decimal CalculateBill(
        decimal itemTotal,
        bool hasPackaging,
        bool hasDelivery,
        bool hasPlatformFee,
        bool hasSurge)
    {
        var total = itemTotal;
        Console.WriteLine($"  [Before] Item total: ₹{total}");

        if (hasPackaging)                               // flag #1
        {
            total += 15;
            Console.WriteLine($"  [Before] + Packaging: ₹15 -> ₹{total}");
        }
        if (hasDelivery)                                // flag #2
        {
            total += 40;
            Console.WriteLine($"  [Before] + Delivery: ₹40 -> ₹{total}");
        }
        if (hasPlatformFee)                             // flag #3
        {
            total += 5;
            Console.WriteLine($"  [Before] + Platform fee: ₹5 -> ₹{total}");
        }
        if (hasSurge)                                   // flag #4
        {
            total += 20;
            Console.WriteLine($"  [Before] + Surge: ₹20 -> ₹{total}");
        }
        // New charge = new flag = new parameter = update every caller
        // Testing? 2^4 = 16 combinations. Add one more flag? 32.

        return total;
    }
}

File: typescript/src/04-decorator/before/bill-calculator.ts
// Boolean flags for each charge; every new fee adds another if-block
export class BillCalculator {
  calculateBill(
    itemTotal: number,
    hasPackaging: boolean,
    hasDelivery: boolean,
    hasPlatformFee: boolean,
    hasSurge: boolean,
    hasConvenienceFee = false,
  ): number {
    let total = itemTotal;
    console.log(`  [Before] Item total: ₹${total}`);

    if (hasPackaging) {
      total += 15;
      console.log(`  [Before] + Packaging: ₹15 -> ₹${total}`);
    }
    if (hasDelivery) {
      total += 40;
      console.log(`  [Before] + Delivery: ₹40 -> ₹${total}`);
    }
    if (hasPlatformFee) {
      total += 5;
      console.log(`  [Before] + Platform fee: ₹5 -> ₹${total}`);
    }
    if (hasSurge) {
      total += 20;
      console.log(`  [Before] + Surge: ₹20 -> ₹${total}`);
    }
    if (hasConvenienceFee) {
      total += 5;
      console.log(`  [Before] + Convenience fee: ₹5 -> ₹${total}`);
    }
    // Testing? 2^5 = 32 combinations. Overlap: platform + convenience both add fee.

    return total;
  }
}